### PR TITLE
Javadoc updates for FuturesExtra

### PR DESCRIPTION
**Pull request by noaresare**

### CODE REVIEW

File: src/main/java/com/spotify/futures/FuturesExtra.java [hunk 1]
```diff
@@ -153,6 +158,18 @@ public void onFailure(Throwable t) {
     return promise;
   }
 
+  /**
+   * Transform the results of two {@link ListenableFuture}s a and b once both has completed using
+   * {@code function} and complete the returned ListenableFuture at that point. This method is
+   * synchronous in the sense that function.apply() executes synchronously and returns a value
+   * and not a {@link java.util.concurrent.Future}.
```
Comment (spkrka) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: This should reference ListenableFuture instead of Future

File: src/main/java/com/spotify/futures/FuturesExtra.java [hunk 2]
```diff
@@ -323,6 +474,23 @@ public Z apply(List<Object> results) {
...
+   *
+   * @param a the first ListenableFuture
+   * @param b the second ListenableFuture
+   * @param c the third ListenableFuture
+   * @param d the fourth ListenableFuture
+   * @param e the fifth ListenableFuture
+   * @param f the sixth ListenableFuture
+   * @param function an AsyncFunction6 implementation, possibly expressed as a java 8 lambda.
+   * @return a ListenableFuture of the result of {@code function}, that completes as soon as
+   * the future that gets returned by the supplied Function6.apply() has completed.
```
Comment (spkrka) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: this is not technically true - you could return a future that be completed before the transform has even started

**Comment by spkrka** [comment]
Can't we just document the (a?)syncTransform[2-6] functions with something like this:
"This is equivalent to Futures.transform(Futures.allAsList(f1, f2, ...))"

Or maybe copy some text from here: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/util/concurrent/Futures.html#transform(com.google.common.util.concurrent.ListenableFuture,%20com.google.common.util.concurrent.AsyncFunction)

"Returns a new ListenableFuture whose result is asynchronously derived from the result of the given tuple of Futures. More precisely, the returned Future takes its result from the Future produced by applying the given AsyncFunction to the result of the original Futures"

File: src/main/java/com/spotify/futures/FuturesExtra.java (hunk 2 above)
Comment (noaresare) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: I'm not sure I follow. As far as I can tell, there is no way that the ListenableFuture<Z> returned by asyncTransform6 can complete without the ListenableFuture<Z> returned by the supplied AsyncFunction6 implementation has completed. Sure, I guess you can have an apply() that returns a ListenableFuture that completes immediately and then do a "transform" in a separate thread somewhere without the result of that transform being directly available in the code but would argue that that stretches the definition of transform a bit.

Or, as always, I could be totally confused, in which case I'm happy to be directed in the right direction :)

Comment (spkrka) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: Example:
LF complete = immediateFuture("complete");
LF result = FuturesExtra.asyncTransform2(futureA, futureB, (a, b) -> complete);

result is NOT completed as soon as the future returned by the lambda is completed.
I think we should avoid specifying the exact semantics - it's much easier to describe it as a wrapper around the actual transform function.

Comment (noaresare) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: Oh, now I see. Should have thought of that. I toyed around with a wording that actually referenced the guava functionality that this group of methods wrap, but it became hard to read (you would typically need to be acquainted with the idiom of putting them in an allAsList future to be able to explain it). I think that well written javadoc should help people reading the code in the future to abstract away somewhat complex things.

Unfortunately the wording of upstream doesn't really apply. What do you think about the following wording?
A future that holds result of the function (if all inputs succeeded) or the first failing input's failure (if not)

**Comment by noaresare** [comment]
I managed to miss this comment when responding to the other one. Seeing how many errors I made when copy/pasting the various variants with updates for number of parameters etc I think that we should at least make the wording number of argument-neutral.

File: src/main/java/com/spotify/futures/FuturesExtra.java [hunk 3]
```diff
@@ -153,6 +158,18 @@ public void onFailure(Throwable t) {
...
 
+  /**
+   * Transform the results of two {@link ListenableFuture}s a and b once both has completed using
+   * {@code function} and complete the returned ListenableFuture at that point. This method is
+   * synchronous in the sense that function.apply() executes synchronously and returns a value
+   * and not a {@link ListenableFuture}.
+   *
+   * @param a the first ListenableFuture
+   * @param b the second ListenableFuture
+   * @param function a Function2 implementation, possibly expressed as a java 8 lambda
```
Comment (spkrka) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: "possibly expressed as a java 8 lambda" is not really relevant

File: src/main/java/com/spotify/futures/FuturesExtra.java [hunk 4]
```diff
@@ -153,6 +158,18 @@ public void onFailure(Throwable t) {
     return promise;
   }
 
+  /**
+   * Transform the results of two {@link ListenableFuture}s a and b once both has completed using
+   * {@code function} and complete the returned ListenableFuture at that point. This method is
+   * synchronous in the sense that function.apply() executes synchronously and returns a value
+   * and not a {@link ListenableFuture}.
```
Comment (spkrka) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: I would rephrase all this as:

Transform the two input futures (a and b) into a single future, using the provided transform function. The transformation follows the same semantics as as Futures.transform and the two input futures are combined using Futures.allAsList.

File: src/main/java/com/spotify/futures/FuturesExtra.java [hunk 5]
```diff
@@ -153,6 +158,18 @@ public void onFailure(Throwable t) {
...
+   *
+   * @param a the first ListenableFuture
+   * @param b the second ListenableFuture
+   * @param function a Function2 implementation, possibly expressed as a java 8 lambda
+   * @return a ListenableFuture of the result of {@code function}, that completes as soon as
+   * the supplied Function2.apply() invocation has completed.
```
Comment (spkrka) [inline src/main/java/com/spotify/futures/FuturesExtra.java]: no need to mention anything about completing futures.

**Comment by noaresare** [comment]
I agree with all your points. I have now created a simple generic input count neutral javadoc for the transform family of functions. The only difference between the sync and async variant is which variant of Futures.transform() it references. Thank you for your patience.

**Comment by spkrka** [comment]
Thanks for taking the time to write documentation :)

:+1:
